File: app/src/main/java/com/hiddenramblings/tagmo/amiibo/AmiiboComparator.java
package com.hiddenramblings.tagmo.amiibo;

import com.hiddenramblings.tagmo.settings.BrowserSettings;
import com.hiddenramblings.tagmo.settings.BrowserSettings.SORT;

import java.io.File;
import java.util.Comparator;

public class AmiiboComparator implements Comparator<AmiiboFile> {

    BrowserSettings settings;

    public AmiiboComparator(BrowserSettings settings) {
        this.settings = settings;
    }

    @Override
    public int compare(AmiiboFile amiiboFile1, AmiiboFile amiiboFile2) {
        int value = 0;
        int sort = settings.getSort();

        File filePath1 = amiiboFile1.getFilePath();
        File filePath2 = amiiboFile2.getFilePath();
        if (sort == SORT.FILE_PATH.getValue() && !(null == filePath1  && null == filePath2 ))
            value = compareFilePath(filePath1, filePath2);

        long amiiboId1 = amiiboFile1.getId();
        long amiiboId2 = amiiboFile2.getId();
        if (sort == SORT.ID.getValue()) {
            value = compareAmiiboId(amiiboId1, amiiboId2);
        } else if (value == 0) {
            AmiiboManager amiiboManager = settings.getAmiiboManager();
            if (null != amiiboManager ) {
                Amiibo amiibo1 = amiiboManager.amiibos.get(amiiboId1);
                Amiibo amiibo2 = amiiboManager.amiibos.get(amiiboId2);
                if (null == amiibo1  && null == amiibo2 )
                    //noinspection all
                    value = 0;
                else if (null == amiibo1 )
                    value = 1;
                else if (null == amiibo2 )
                    value = -1;
                else if (sort == SORT.NAME.getValue()) {
                    value = compareAmiiboName(amiibo1, amiibo2);
                } else if (sort == SORT.AMIIBO_SERIES.getValue()) {
                    value = compareAmiiboSeries(amiibo1, amiibo2);
                } else if (sort == SORT.AMIIBO_TYPE.getValue()) {
                    value = compareAmiiboType(amiibo1, amiibo2);
                } else if (sort == SORT.GAME_SERIES.getValue()) {
                    value = compareGameSeries(amiibo1, amiibo2);
                } else if (sort == SORT.CHARACTER.getValue()) {
                    value = compareCharacter(amiibo1, amiibo2);
                }
                if (value == 0 && null != amiibo1 )
                    value = amiibo1.compareTo(amiibo2);
            }
            if (value == 0)
                value = compareAmiiboId(amiiboId1, amiiboId2);
        }

        if (value == 0)
            value = compareFilePath(filePath1, filePath2);

        return value;
    }

    int compareFilePath(File filePath1, File filePath2) {
        if (null == filePath1 ) {
            return 1;
        } else if (null == filePath2 ) {
            return -1;
        } else {
            return filePath1.compareTo(filePath2);
        }
    }

    int compareAmiiboId(long amiiboId1, long amiiboId2) {
        return Long.compare(amiiboId1, amiiboId2);
    }

    int compareAmiiboName(Amiibo amiibo1, Amiibo amiibo2) {
        String name1 = amiibo1.name;
        String name2 = amiibo2.name;
        if (null == name1  && null == name2 ) {
            return 0;
        }
        if (null == name1 ) {
            return 1;
        } else if (null == name2 ) {
            return -1;
        }
        return name1.compareTo(name2);
    }

    int compareAmiiboSeries(Amiibo amiibo1, Amiibo amiibo2) {
        AmiiboSeries amiiboSeries1 = amiibo1.getAmiiboSeries();
        AmiiboSeries amiiboSeries2 = amiibo2.getAmiiboSeries();
        if (null == amiiboSeries1  && null == amiiboSeries2  ) {
            return 0;
        }
        if (null == amiiboSeries1 ) {
            return 1;
        } else if (null == amiiboSeries2) {
            return -1;
        }
        return amiiboSeries1.compareTo(amiiboSeries2);
    }

    int compareAmiiboType(Amiibo amiibo1, Amiibo amiibo2) {
        AmiiboType amiiboType1 = amiibo1.getAmiiboType();
        AmiiboType amiiboType2 = amiibo2.getAmiiboType();
        if (null == amiiboType1 && null == amiiboType2 ) {
            return 0;
        }
        if (null == amiiboType1) {
            return 1;
        } else if (null == amiiboType2) {
            return -1;
        }
        return amiiboType1.compareTo(amiiboType2);
    }

    int compareGameSeries(Amiibo amiibo1, Amiibo amiibo2) {
        GameSeries gameSeries1 = amiibo1.getGameSeries();
        GameSeries gameSeries2 = amiibo2.getGameSeries();
        if (null == gameSeries1 && null == gameSeries2) {
            return 0;
        }
        if (null == gameSeries1) {
            return 1;
        } else if (null == gameSeries2) {
            return -1;
        }
        return gameSeries1.compareTo(gameSeries2);
    }

    int compareCharacter(Amiibo amiibo1, Amiibo amiibo2) {
        Character character1 = amiibo1.getCharacter();
        Character character2 = amiibo2.getCharacter();
        if (null == character1 && null == character2) {
            return 0;
        }
        if (null == character1) {
            return 1;
        } else if (null == character2) {
            return -1;
        }
        return character1.compareTo(character2);
    }
}

File: app/src/main/java/com/hiddenramblings/tagmo/amiibo/AmiiboComparator.kt
package com.hiddenramblings.tagmo.amiibo

import com.hiddenramblings.tagmo.BrowserSettings
import com.hiddenramblings.tagmo.BrowserSettings.SORT

class AmiiboComparator(var settings: BrowserSettings) : Comparator<Amiibo> {
    override fun compare(amiiboFile1: Amiibo, amiiboFile2: Amiibo): Int {
        var value = 0
        val sort = settings.sort
        val amiiboId1 = amiiboFile1.id
        val amiiboId2 = amiiboFile2.id
        if (sort == SORT.FILE_PATH.value || sort == SORT.ID.value) {
            value = compareAmiiboId(amiiboId1, amiiboId2)
        } else {
            settings.amiiboManager?.let {
                val amiibo1 = it.amiibos[amiiboId1]
                val amiibo2 = it.amiibos[amiiboId2]
                value = when {
                    null == amiibo1 && null == amiibo2 -> 0
                    null == amiibo1 -> 1
                    null == amiibo2 -> -1
                    sort == SORT.NAME.value -> compareAmiiboName(amiibo1, amiibo2)
                    sort == SORT.AMIIBO_SERIES.value -> compareAmiiboSeries(amiibo1, amiibo2)
                    sort == SORT.AMIIBO_TYPE.value -> compareAmiiboType(amiibo1, amiibo2)
                    sort == SORT.GAME_SERIES.value -> compareGameSeries(amiibo1, amiibo2)
                    sort == SORT.CHARACTER.value -> compareCharacter(amiibo1, amiibo2)
                    else -> 0
                }
            }
            if (value == 0) value = compareAmiiboId(amiiboId1, amiiboId2)
        }
        return value
    }

    private fun compareAmiiboId(amiiboId1: Long, amiiboId2: Long): Int {
        return amiiboId1.compareTo(amiiboId2)
    }

    private fun compareAmiiboName(amiibo1: Amiibo, amiibo2: Amiibo): Int {
        val name1 = amiibo1.name
        val name2 = amiibo2.name
        return when {
            null == name1 && null == name2 -> 0
            null == name1 -> 1
            null == name2 -> -1
            else -> name1.compareTo(name2)
        }
    }

    private fun compareAmiiboSeries(amiibo1: Amiibo, amiibo2: Amiibo): Int {
        val amiiboSeries1 = amiibo1.amiiboSeries
        val amiiboSeries2 = amiibo2.amiiboSeries
        return when {
            null == amiiboSeries1 && null == amiiboSeries2 -> 0
            null == amiiboSeries1 -> 1
            null == amiiboSeries2 -> -1
            else -> amiiboSeries1.compareTo(amiiboSeries2)
        }
    }

    private fun compareAmiiboType(amiibo1: Amiibo, amiibo2: Amiibo): Int {
        val amiiboType1 = amiibo1.amiiboType
        val amiiboType2 = amiibo2.amiiboType
        return when {
            null == amiiboType1 && null == amiiboType2 -> 0
            null == amiiboType1 -> 1
            null == amiiboType2 -> -1
            else -> amiiboType1.compareTo(amiiboType2)
        }
    }

    private fun compareGameSeries(amiibo1: Amiibo, amiibo2: Amiibo): Int {
        val gameSeries1 = amiibo1.gameSeries
        val gameSeries2 = amiibo2.gameSeries
        return when {
            null == gameSeries1 && null == gameSeries2 -> 0
            null == gameSeries1 -> 1
            null == gameSeries2 -> -1
            else -> gameSeries1.compareTo(gameSeries2)
        }
    }

    private fun compareCharacter(amiibo1: Amiibo, amiibo2: Amiibo): Int {
        val character1 = amiibo1.character
        val character2 = amiibo2.character
        return when {
            null == character1 && null == character2 -> 0
            null == character1 -> 1
            null == character2 -> -1
            else -> character1.compareTo(character2)
        }
    }
}